implementation/src/main/java/io/smallrye/mutiny/operators/UniMemoizeOp.java

```diff
@@ -0,0 +1,181 @@
+
+            if (!awaitingSubscription.isEmpty()) {
+                // Make a safe copy of the subscribers awaiting for a subscription
+                synchronized (awaitingSubscription) {
+                    subscribers = new ArrayList<>(awaitingSubscription);
+                }
+
+                // Handle the subscribers that are awaiting a subscription
+                for (UniSerializedSubscriber<? super I> subscriber : subscribers) {
+                    currentItem = item;
+                    currentFailure = failure;
+                    State state = this.state.get();
+                    switch (state) {
+                        case INIT:
+                        case SUBSCRIBING:
+                            break;
+                        case SUBSCRIBED:
+                            subscriber.onSubscribe(() -> tryCancelSubscription(subscriber));
+                            awaitingSubscription.remove(subscriber);
+                            awaitingResult.add(subscriber);
+                            break;
+                        case CACHING:
+                            subscriber.onSubscribe(() -> tryCancelSubscription(subscriber));
+                            if (currentFailure != null) {
+                                subscriber.onFailure(currentFailure);
+                            } else {
+                                subscriber.onItem(currentItem);
+                            }
+                            awaitingSubscription.remove(subscriber);
+                            awaitingResult.remove(subscriber);
+                            break;
+                        default:
+                            throw new IllegalStateException("Current state is " + state);
+                    }
+                }
+            }
+
+            if (!awaitingResult.isEmpty()) {
+                // Make a safe copy of the subscribers awaiting a result
+                synchronized (awaitingResult) {
+                    subscribers = new ArrayList<>(awaitingResult);
+                }
+                // Handle the subscribers that are awaiting a result
+                for (UniSerializedSubscriber<? super I> subscriber : subscribers) {
+                    currentItem = item;
+                    currentFailure = failure;
+                    if (state.get() == State.CACHING) {
+                        if (failure != null) {
+                            subscriber.onFailure(currentFailure);
+                        } else {
+                            subscriber.onItem(currentItem);
+                        }
+                        awaitingResult.remove(subscriber);
+                    }
+                }
+            }
+
+            missed = wip.addAndGet(-missed);
+            if (missed == 0) {
+                break;
+            }
+        }
+    }
+
+    private void tryCancelSubscription(UniSerializedSubscriber<? super I> subscriber) {
+        awaitingSubscription.remove(subscriber);
```

Comment:
Misleading name no? It does not cancel anything.

Reply:
No but it removes the subscriber from notification lists.

Reply:
Renaming